### PR TITLE
Initial support for the ESP32-C3 early dev board

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

The [ESP32-C3](https://www.espressif.com/en/news/ESP32_C3) is a MCU from Espressif that uses a custom ISA. The ISA is based on RISC-V and shares lots of features and functionality but does not follow the RISC-V spec. It is close enough though that we can run RISC-V Tock on the core.

This PR adds support for the ESP32-C3-DevKitM-1 v1.6 board.

### Testing Strategy

I have tested this on an engineering sample board that I have.

### TODO or Help Wanted

- [x] GPIO support

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by alistair23** [comment]
This is ready to go now

File: .gitignore [hunk 1]
```diff
@@ -71,3 +71,6 @@ boards/swervolf/binary.hex
 boards/swervolf/build/
 boards/swervolf/fusesoc.conf
 boards/swervolf/fusesoc_libraries/
+
+# ESP C3 temp hez
```
Comment (bradjc) [inline .gitignore:75]:
```suggestion
# ESP C3 temp hex
```

File: chips/esp32/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+//! Drivers and chip support for Espressif ESP32 boards.
+
+#![feature(asm, const_fn_trait_bound, naked_functions)]
```
Comment (bradjc) [inline chips/esp32/src/lib.rs:3]:
```suggestion
#![feature(const_fn_trait_bound)]
```

File: chips/esp32-c3/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+//! Drivers and chip support for ESP32-C3.
+
+#![feature(asm, const_fn_trait_bound, naked_functions)]
```
Comment (bradjc) [inline chips/esp32-c3/src/lib.rs:3]:
```suggestion
#![feature(const_fn_trait_bound)]
```

File: chips/esp32/src/timg.rs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    fn disarm(&self) {
+        panic!("Unimplemented");
+    }
+
+    fn get_remaining_us(&self) -> Option<u32> {
+        panic!("Unimplemented");
+    }
+}
```
Comment (bradjc) [inline chips/esp32/src/timg.rs:232]: Do we need this?

Comment (hudson-ayers) [reply chips/esp32/src/timg.rs:232]: I would recommend using the virtual scheduler timer instead of this approach

**Comment by alistair23** [comment]
Updated to remove the `unimplemented` functions

**Comment by alistair23** [comment]
Ping

**Comment by phil-levis** [comment]
This looks cool. IIRC, the minimum bar for including a new board in the tree is UART/Console, Alarm, and GPIO capsules/system calls. Could you add GPIO?

**Comment by alistair23** [comment]
Strange, the GPIO commit was dropped. Just pushed it back

File: chips/esp32/src/gpio.rs [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    fn configuration(&self) -> gpio::Configuration {
+        if self.registers.enable.is_set(self.pin) {
+            gpio::Configuration::Input
+        } else {
+            gpio::Configuration::InputOutput
+        }
+    }
+
+    fn set_floating_state(&self, _mode: gpio::FloatingState) {
+        unimplemented!()
```
Comment (phil-levis) [inline chips/esp32/src/gpio.rs:141]: Please add support for pull-up/pull-down.

File: chips/esp32/src/gpio.rs [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    ) -> GpioPin<'a> {
+        GpioPin {
+            registers: gpio_base,
+            pin,
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        unimplemented!()
```
Comment (phil-levis) [inline chips/esp32/src/gpio.rs:127]: Please add support for interrupt dispatch.

**Review by phil-levis — CHANGES_REQUESTED** [review]

**Review by phil-levis — APPROVED** [review]
Looks good!

**Pull request merged**